### PR TITLE
feat: add metadata table cleanup

### PR DESCRIPTION
**Pull request by zoercai**

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 1]
```diff
@@ -1503,13 +1505,17 @@ public ReadChangeStream withTraceSampler(Sampler traceSampler) {
...
-        // TODO: We need to perform cleanup after everything has terminated (delete metadata table)
+        PCollection<DataChangeRecord> results =
+            input
+                .apply("Generate change stream sources", Create.of(sources))
+                .apply("Detect new partitions", ParDo.of(detectNewPartitionsDoFn))
+                .apply("Read change stream partition", ParDo.of(readChangeStreamPartitionDoFn))
+                .apply("Post processing metrics", ParDo.of(postProcessingMetricsDoFn));
+        input
+            .apply(Impulse.create())
+            .apply(Wait.on(results))
```
Comment (hengfengli) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1516]: If I cancel a running pipeline, would it also clean up the metadata table?

Comment (hengfengli) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1516]: Where is this `Wait` symbol from? I couldn't see any import for `Wait`.

**Review by hengfengli — APPROVED** [review]
LGTM.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 1 above)
Comment (hengfengli) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1516]: From my understanding, it should look like this: 

```
PCollection<byte[]> impulseOut = p.apply(Impulse.create());

PCollection<DataChangeRecord> results =
    impulseOut
        .apply("Generate change stream sources", Create.of(sources))       // This step needs to be updated. 
        .apply("Detect new partitions", ParDo.of(detectNewPartitionsDoFn))
        .apply("Read change stream partition", ParDo.of(readChangeStreamPartitionDoFn))
        .apply("Post processing metrics", ParDo.of(postProcessingMetricsDoFn));

impulseOut
        .apply(Wait.on(results))
        .apply(ParDo.of(new CleanUpReadChangeStreamDoFn(daoFactory)));
```

**Review by hengfengli — CHANGES_REQUESTED** [review]
Sorry, I gave it a second thought and think the use of `Impulse.create()` may need to be changed. Probably need to check it with Luke.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 1 above)
Comment (zoercai) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1516]: > If I cancel a running pipeline, would it also clean up the metadata table?

Luke said the only way for the table to be cleaned up if the pipeline is interrupted is to use Spanner TTL

Comment (zoercai) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1521]: > Where is this Wait symbol from? I couldn't see any import for Wait.

It's on line 87, already imported: https://github.com/thiagotnunes/beam/pull/52/files#diff-12d8841cc4fa9802d7b0640b8354aee6d668a2f92fad6491901a216f92cd2ee8R87

**Review by hengfengli — APPROVED** [review]
Thanks for the change. LGTM.

**Pull request merged**